src/test/fidelity/common.ts

```diff
@@ -0,0 +1,223 @@
+  outputDirectory: string;
+  scenarioDirectory: string;
+  analysisThresholds: Array<number>;
+  scenarios: Array<ScenarioConfig>;
+}
+
+export interface ComparableImage {
+  [index: number]: number;
+  length: number;
+}
+
+export interface AnalysisOptions {
+  generateVisuals: boolean;
+}
+
+export class ImageComparator {
+  protected imagePixels: number;
+
+  constructor(
+      protected candidateImage: ComparableImage,
+      protected goldenImage: ComparableImage, readonly dimensions: Dimensions) {
+    const {width, height} = dimensions;
+
+    this.imagePixels = width * height;
+  }
+
+  protected drawPixel(
+      image: ComparableImage, position: number, r: number, g: number, b: number,
+      a: number = 255) {
+    image[position + 0] = r;
+    image[position + 1] = g;
+    image[position + 2] = b;
+    image[position + 3] = a;
+  }
+
+  analyze(threshold: number, options: AnalysisOptions = {
+    generateVisuals: true
+  }): ImageComparisonResults {
+    const {candidateImage, goldenImage} = this;
+    const {width, height} = this.dimensions;
+    const {generateVisuals} = options;
+
+    const booleanImage = generateVisuals ?
+        new Uint8ClampedArray(this.imagePixels * COMPONENTS_PER_PIXEL) :
+        null;
+    const deltaImage = generateVisuals ?
+        new Uint8ClampedArray(this.imagePixels * COMPONENTS_PER_PIXEL) :
+        null;
+
+    const thresholdSquared = threshold * threshold;
+
+    let matched = 0;
+    let sum = 0;
+    let mismatchingSum = 0;
+    let maximumDeltaIntensity = 0;
+
+    if (candidateImage.length != goldenImage.length) {
+      throw new Error(`Image sizes do not match (candidate: ${
+          candidateImage.length}, golden: ${goldenImage.length})`);
+    }
+
+    for (let y = 0; y < height; ++y) {
+      for (let x = 0; x < width; ++x) {
+        const index = y * width + x;
+        const position = index * 4;
+        const delta =
+            colorDelta(candidateImage, goldenImage, position, position);
+        const bool = (delta < thresholdSquared ? 1 : 0) * 255;
+
+        if (bool > 0) {
+          matched++;
+        } else {
+          mismatchingSum += delta;
+        }
+
+        const thresholdDelta = Math.max(0, delta - thresholdSquared);
+
+        sum += thresholdDelta;
+
+        if (generateVisuals) {
+          const deltaIntensity =
+              Math.round(255 * thresholdDelta / MAX_COLOR_DISTANCE);
+
+          maximumDeltaIntensity =
+              Math.max(deltaIntensity, maximumDeltaIntensity);
+
+          this.drawPixel(booleanImage!, position, bool, bool, bool);
+          this.drawPixel(
+              deltaImage!,
+              position,
+              255,
+              255 - deltaIntensity,
+              255 - deltaIntensity);
+        }
+      }
+    }
+
+    if (generateVisuals) {
+      for (let y = 0; y < height; ++y) {
+        for (let x = 0; x < width; ++x) {
+          const index = y * width + x;
+          const position = index * 4;
```

Comment:
I don't think COMPONENTS_PER_PIXEL should be a constant (IMO one of the few things that can get away with being a magic number in some contexts, e.g. 255) but should be consistent

Reply:
Something to think about, but implicit meaning is almost always worse. `255` at least as the context that a `Uint8ClampedArray` only holds values in the range 0..255.  `4` can't even say that much.

Reply:
[go/tott/538](http://go/tott/538)